java-checks-test-sources/src/main/java/checks/RecordInsteadOfClassCheck.java

```diff
@@ -0,0 +1,101 @@
+package checks;
+
+public class RecordInsteadOfClassCheck {
+  class SimpleClass implements NotAClass { // Noncompliant [[sc=9;ec=20]] {{Refactor this class declaration to use 'record SimpleClass(int sum)'.}}
+    private final int sum;
+    public static final int VALUE = 42;
+
+    SimpleClass(int sum) { this.sum = sum; }
+    int getSum() { return sum; }
+    @Override public void foo() { }
+  }
+
+  class ComplexClass { // Noncompliant {{Refactor this class declaration to use 'record ComplexClass(int sum, List<...> list, boolean[][] tests, boolea...)'.}}
+    private final int sum;
+    private final java.util.List<String> list;
+    private final boolean[][] tests;
+    private final ComplexClass[] classes;
+    private final boolean colorBlind;
+
+    ComplexClass(int sum, java.util.List<String> list, boolean[][] tests, boolean colorBlind, ComplexClass ... classes) {
+      this.sum = sum;
+      this.list = list;
+      this.tests = tests;
+      this.classes = classes;
+      this.colorBlind = colorBlind;
+    }
+
+    int getSum() { return sum; }
+    java.util.List<String> getList() { return list; }
+    boolean[][] getTests() { return tests; }
+    ComplexClass[] getClasses() { return classes; }
+    boolean isColorBlind() { return colorBlind; }
+  }
+
+  class SimpleWithoutGettersForAllFields {
+    private final int sum;
+    private final int base;
+
+    SimpleWithoutGettersForAllFields(int sum, int base) { this.sum = sum; this.base = base; }
+    int getSum() { return sum; }
+  }
+
+  class SimpleWithConstructorNotSettingAllParameters {
+    private final int sum;
+    private final int base;
+
+    SimpleWithConstructorNotSettingAllParameters(int sum) { this.sum = sum; base = 42; }
+    int getSum() { return sum; }
+    int getBase() { return base; }
+  }
+
+  class SimpleWithConstructorNotSettingAllParameters2 {
+    private final int sum;
+    private final int base;
+
+    SimpleWithConstructorNotSettingAllParameters2(int sum, String other) { this.sum = sum; base = 42; }
+    int getSum() { return sum; }
+    int getBase() { return base; }
+  }
+
+
+  class ClassWithSetter {
+    private final int sum;
+
+    ClassWithSetter(int sum) { this.sum = sum; }
+    int getSum() { return sum; }
+    void setSum(int value) { }
```

Comment:
I'm not sure it makes sense to me that we make an exception if and only if there's a method that exactly matches the typical signature for a setter. Is this about the possibility of using reflection to set the field? I don't see the use case where a user would define the fields as final, then define normal-looking setter methods and use reflection inside them to get rid of the finality.

Reply:
I'm also a bit puzzled by this condition. If fields are final and private, setters do not make sense. Probably we can drop the condition from the rule